app/users.go

```diff
@@ -26,66 +26,80 @@ type User struct {
 	Password  string
 }
 
-type CreateUserRequest struct {
-	FirstName string
-	LastName  string
-	Email     string
-	Username  string
-	Password  string
+type RegistrationRequest struct {
+	ErrorFirstName string
+	ErrorLastName  string
+	ErrorEmail     string
+	ErrorUsername  string
+	ErrorPassword  string
+	FirstName      string
+	LastName       string
+	Email          string
+	Username       string
+	Password       string
 }
 
-type FieldError struct {
-	Message string
+type PasswordResetRequest struct {
+	UserId        int
+	Password      string
+	ErrorPassword string
 }
 
-func (e *FieldError) Error() string {
-	return e.Message
-}
+func (s *UserService) validateRegistrationFields(req *RegistrationRequest) bool {
+	v := &validation.Validator{}
+	v.Errors = make(map[string][]error)
+	v.IsEmpty("First name", req.FirstName)
+	v.IsEmpty("Last name", req.LastName)
+	v.IsEmpty("Username", req.Username)
+	v.IsEmpty("Email", req.Email)
+	v.IsEmpty("Password", req.Password)
+	v.IsEmail("Email", req.Email)
+	v.IsValidPassword("Password", req.Password)
+	s.isEmailUsed(v, req.Email)
 
-func (s *UserService) Create(req CreateUserRequest) (*User, error) {
-	ErrorMessage := fieldsNotEmpty(req)
-	if ErrorMessage != "" {
-		return nil, &FieldError{Message: ErrorMessage}
-	}
+	return req.setRegistrationFieldErrors(v.Errors)
```

Comment:
I'll place the comment here because it's a good starting point for what I'm about to say, but it refers to a larger issue.

The complaint I have about this implementation is this writing of errors into the request object. You added error fields into the (now called) `RegistrationRequest` struct, which in my opinion may work in this small example, but is generally harder to maintain. I'll explain my reasoning behind this and offer a potentially better alternative.

When you program anything, there's a practice of generally avoiding mutating/changing data if it's not required. You want to have control over application state at every given moment and to achieve that you want to limit mutations to predictable places in you code. What this implies is that if I have a parameter that I send into the function, I would generally not expect the function to mutate that parameter. Especially this case, where I send a struct of fields that are used to create a new user - you're expecting the caller of the function to use that struct to handle errors. The purpose of this struct is only to group data required to create the user and to implement validations for those fields and nothing else. So that's the first thing we need to clear - avoid using input parameters as output parameters, especially in this case because you can return errors from the function instead of writing them to the struct. Sometimes you want and need output parameters, but this is not the case.

What I would suggest as a better alternative is returning errors using the `error` interface that is intended for this purpose. Because `error` is an interface, you can have any struct (any type, really) implement this interface by implementing the `Error() string` function/method. If you take a look at the `validation.Validator` struct, you'll see that you already implement this `error` interface for that type. Plus, you're already storing the mapping of field names to errors in that `Validator` struct. So what can we do about returning this error properly?

Instead of having the `Validator` struct implement the `error` interface, you can create a new struct in the same package called `Error` (it's fine to call it just `Error` because it's in the `validaiton` pacakge, so it's full name would be `validation.Error`). This `Error` struct would have the same `Errors` field that the validator has. On the `Validator`, have the `Error` function return the `Error` struct like this:

```go
type Validator struct {
  errors map[string][]errors // NOTE: errors is private
}

func New() *Validator {
  return &Validator{
    errors: map[string][]errors{}, // you initialize the map here, instead of the request validation method
  }
}

type Error struct {
    Errors map[string][]error // NOTE: Errors is public
}

func (e *Error) Error() string {
  return "validation error"
}

func (v *Validator)  Validate() *Error {
  // TODO: if map is empty, return nil here to indicate that there are no errors
  return &Error{ errors: v.errors } // this works because Error implements the `error` interface
}
```

Now what this enables you to do is have the validation method on the `RegistrationRequest` return an `error` instead of a bool, like this:
```go
func (r *RegistrationRequest) validate() error {
  v := validator.New()
  // do your validation methods here...
  return v.Validate() // v.Validate() returns Error, which implements `error`
}
```

In turn, when you want to validate the request in the `UserService.Create` method, you can do this:
```go
func (s *UserService) Create(req CreateUserRequest) (*User, error) {
  if err := req.validate(); err != nil {
    return nil, err
  }
  // continue, knowing the req struct is valid...
}
```

In the http handler, when you want to check if it was some other error or a a validation error, you can do a type assertion  or type switch ([example here](https://yourbasic.org/golang/type-assertion-switch/)) to differently handle the errors:
```go
/// ...
user, err := userService.Create(req)

if err != nil {
    switch v := err.(type) {
    case validation.Error:
        // `v` is of the matched type `validation.Error`, so you can access the `.Errors` map
    default:
       // handle other errors as is
    }
}
// ...
```

Now, if you remember what I talked about when passing data to the template renderer, you could create the following:
```go
type Page struct {
  Success bool // this is optional, but maybe useful
  FormFields map[string]string // field name => field value from request
  FormErrors map[string]string // field name => errors
}
```

So you can populate this `Page.FormErrors` from the above `validator.Error.Errors` map.

Also, to handle this User unique email check, you could implement it so that the `validation.Validator` has a method do just add any error to the field:
```go
// something like this...
func (v *Validator) AddError(fieldName string, err error) {
  v.errors[fieldName] = append(v.errors[fieldName], err)
}
```

Now, you would add the validation logic to you `RegistrationRequest.validate` method like this:
```go
// NOTE: I added the UserService as a parameter, so I can use it to check if a user with the given email exists...
func (r *RegistrationRequest) validate(s *UserService) error {
  v := validator.New()
  // do your validation methods here...
  
  // do a call on the service to check if a user exists with the given email
  if s.hasUserWithEmail(r.Email) {
    // add the error to the "email" field or just return the `error` from `fmt.Errorf` directly; I prefer to add it to the "email" field
    v.AddError("email", fmt.Errorf("email is not unique or whatever message")
  }
  
  return v.Validate() // v.Validate() returns Error, which implements `error`
}
```
I'm sorry for the lengthy comment, but I really wanted to clear this up, because there are multiple things that I wanted addressed and I really think you shouldn't go about error handling with the direction you went. I hope what I suggested made it a bit clearer how the error handling could be done. If you have more questions please post them and I'll try to respond as soon as I can. I don't want you to blindly copy/paste the solution, but I would like you to go in this direction with understanding.

I want you both to learn from this about how you can pass things around and what to avoid. It know it might seem complicated, but it's really simple once you see what's actually going on and how you can now apply this validation package and error handling to any other resource you might create, not just the User. You saw how I still managed to keep email uniqueness validation in the `RegistrationRequest.validate` method, instead of scattering logic all about. This passing of dependencies and localizing logic is really useful and powerful to manage complexity once you get used to it. I'm not expecting you to completely learn everything and write perfect code (there's no such thing), but I do expect you to keep improving, so don't let this huge comment discourage you.

P.S, I would really like if you reverted the the name back to `CreateUserRequest` instead of `RegistartionRequest` xD.